[Blog] Reducing AD False Positives Through Algorithmic Improvements (#3524)
opensearch-project/project-website: _posts/2024-12-26-Reducing-false-positives-through-algorithmic-improvements.md

---
layout: post
title: "Enhancing OpenSearch anomaly detection: Reducing false positives through algorithmic improvements"
authors:
  - kaituo
  - kolchfa

> date: 2024-12-26

Reviewer: ```suggestion
date: 2024-02-05
```

> date: 2024-12-26

Reviewer: Done

categories:
  - technical-posts
has_math: true
has_science_table: true

> meta_keywords: anomaly detection, false positives, algorithmic improvements, OpenSearch, nab benchmark

Reviewer 2: Please update the meta with the following: 
_meta_keywords:_ anomaly detection, OpenSearch RCF algorithm, Random Cut Forest, value vectors, Numenta Anomaly Benchmark, algorithmic improvements in OpenSearch

_meta_description:_ Explore how OpenSearch 2.17 enhanced the Random Cut Forest algorithm, making is possible to reduce false positives by 94.3% while maintaining detection accuracy. Get benchmark results and read practical studies.

Reviewer: Done

> meta_keywords: anomaly detection, false positives, algorithmic improvements, OpenSearch, nab benchmark

Reviewer: ```suggestion
meta_keywords: anomaly detection, OpenSearch RCF algorithm, Random Cut Forest, value vectors, Numenta Anomaly Benchmark, algorithmic improvements in OpenSearch
```

> meta_description: Explore how recent algorithmic improvements in RCF reduces false positives in OpenSearch Anomaly Detection. Illustrate the improvements with NAB benchmark.

Reviewer: ```suggestion
meta_description: Explore how OpenSearch 2.17 enhanced the Random Cut Forest algorithm, making is possible to reduce false positives by 94.3% while maintaining detection accuracy. Get benchmark results and read practical studies.
```

- **Expected observations and their running mean**: The model also examines expected (imputed) values, ensuring they continue to cluster around their own running mean. This step helps confirm that the baseline evolves appropriately.

If both actual and expected values remain within their updated acceptable ranges, the algorithm stops flagging these as anomalies.

### Grouped alerting

To handle transitional periods during data drift or a level shift, the algorithm supports grouped alerting. When multiple unusual events occur close together (within one time window, or _shingle_), you'll receive just one alert instead of many. This approach prevents flooding the user with redundant alerts when the system is still adapting to new conditions.

### Using approximate nearest neighbors for anomaly detection

When deciding whether a candidate event is a true anomaly or merely a normal fluctuation, RCF considers approximate nearest neighbors using the following approach:

1. **Build an expected value vector**: RCF creates a “normal” version of the queried point by replacing suspicious features with typical historical values.

> 2. **Find neighbors in each tree**: RCF traverses each tree along its splitting dimensions to find a leaf node close to the expected vector. Closeness is measured using a scaling factor $$C$$ (for example, $$C$$ = 1.1), which determines the threshold for how far neighbors can be from the expected vector. If a sufficient number of trees (for example, at least 10%) detect neighbors within $$C$$ multiplied by the average distance, the event is considered likely normal.

Reviewer 3: ```suggestion
2. **Find neighbors in each tree**: RCF traverses each tree along its splitting dimensions to find a leaf node close to the expected vector. Closeness is measured using a scaling factor $$C$$ (for example, $$C$$ = 1.1), which determines the distance threshold for how far neighbors can be from the expected vector. If a sufficient number of trees (for example, at least 10%) detect neighbors within $$C$$ multiplied by the average distance, the event is considered likely to be normal.
```

> 3. **Determine the “safe box”**: RCF also defines a “safe box” around the original point, based on factors such as recent noise levels and average deviations. If most neighbors lie within this box, the point is deemed “explainable.” If too few neighbors fall inside, the point is flagged as an anomaly.

Reviewer 3: ```suggestion
3. **Determine the "safe box"**: RCF also defines a "safe box" around the original point, based on factors such as recent noise levels and average deviations. If most neighbors lie within this box, the point is deemed "explainable." If too few neighbors fall inside the box, the point is flagged as an anomaly.
```

### Rescoring using expected value vectors

> To further reduce false positives caused by minor fluctuations and during transitional periods (such as data drifts or level shifts), RCF rescales the anomaly score using an expected value vector. Essentially, it replaces potentially anomalous segments with similar historical values in order to compute a new score. If the difference between this “expected” score and the original score remains below a scaled threshold (factoring in the prior anomaly’s score and the current threshold), RCF decides not to raise another alert. This strategy helps avoid repeated notifications for the same anomaly.

Reviewer 3: ```suggestion
To further reduce false positives caused by minor fluctuations and during transitional periods (such as data drifts or level shifts), RCF rescales the anomaly score using an expected value vector. Essentially, it replaces potentially anomalous segments with similar historical values in order to compute a new score. If the difference between this "expected" score and the original score remains below a scaled threshold (factoring in the prior anomaly's score and the current threshold), RCF does not raise another alert. This strategy helps avoid repeated notifications for the same anomaly.
```

## Evaluating RCF using the Numenta Anomaly Benchmark

> To test these improvements, the AD team evaluated RCF using labeled AWS CloudWatch metrics from the [Numenta Anomaly Benchmark (NAB)](https://github.com/numenta/NAB). The datasets include AWS server performance metrics, such as CPU utilization, incoming network traffic (measured in bytes), and disk write activity (measured in bytes). These datasets also contain verified anomaly labels provided by Numenta, which serve as a reliable reference for identifying abnormal patterns.

Reviewer 3: ```suggestion
To test these improvements, the Anomaly Detection team evaluated RCF using labeled Amazon CloudWatch metrics from the [Numenta Anomaly Benchmark (NAB)](https://github.com/numenta/NAB). The datasets include AWS server performance metrics, such as CPU utilization, incoming network traffic (measured in bytes), and disk write activity (measured in bytes). These datasets also contain verified anomaly labels provided by Numenta, which serve as a reliable reference for identifying abnormal patterns.
```

### Test configuration

We used the following RCF configuration for testing:

- 50 trees, each trained on 256 samples  
- A shingle size of 8  
- A warm-up period of 40 data points  
- A constraint requiring actual values to deviate by at least 20% from expected values  

> These settings match the default RCF parameters in the OpenSearch AD plugin.

Reviewer 3: ```suggestion
These settings match the default RCF parameters in the Anomaly Detection plugin.
```

  - 2014-03-10 21:09:00 (Detected)  
- **Total RCF anomalies**: 3  

![disk write bytes](/assets/media/blog-images/2024-12-26-Reducing-false-positives-through-algorithmic-improvements/disk.png){:class="img-centered"}

#### RDS CPU utilization (rds_cpu_utilization_e47b3b)

- **Dataset**: NAB AWS `rds_cpu_utilization_e47b3b`  
- **Total records**: 4,031  
- **Labeled anomalies**:  
  - 2014-04-13 06:52:00 (Detected)  
  - 2014-04-18 23:27:00 (Detected)  
- **Total RCF anomalies**: 2  

![RDS CPU utilization](/assets/media/blog-images/2024-12-26-Reducing-false-positives-through-algorithmic-improvements/rds.png){:class="img-centered"}

#### RDS CPU utilization (rds_cpu_utilization_cc0c53)

- **Dataset**: NAB AWS `rds_cpu_utilization_cc0c53`  
- **Total records**: 4,031  
- **Labeled anomalies**:  
  - 2014-02-25 07:15:00 (Detected)  
  - 2014-02-27 00:50:00 (Undetected)  
- **Total RCF anomalies**: 1  

![rds_cc0c53](/assets/media/blog-images/2024-12-26-Reducing-false-positives-through-algorithmic-improvements/rds_cc.png){:class="img-centered"}

## Precision and recall

Before we present the results, let's define two key concepts in machine learning evaluation:

> - **Precision**: The proportion of detected anomalies that are correct: of all the anomalies the algorithm detected, how many were true anomalies? High precision means few mistaken identifications. For example, if the algorithm detects 10 anomalies and 8 are actually anomalies, then the precision is $$8 / 10 = 0.8$$, or $$80%$$.  

Reviewer 3: ```suggestion
- **Precision**: The proportion of detected anomalies that are correct---of all the anomalies the algorithm detected, how many were true anomalies? High precision results in fewer mistaken identifications. For example, if the algorithm detects 10 anomalies and 8 are actually anomalies, then the precision is $$8 / 10 = 0.8$$, or $$80%$$.  
```

> - **Recall**: The proportion of true anomalies that are detected: of all true anomalies, how many did the algorithm detect? High recall means few missed anomalies. For example, if there are 10 true anomalies and the algorithm detects 8 of them, then the recall is $$8 / 10 = 0.8$$, or $$80%$$.  

Reviewer 3: ```suggestion
- **Recall**: The proportion of true anomalies that are detected---of all true anomalies, how many did the algorithm detect? High recall results in fewer missed anomalies. For example, if there are 10 true anomalies and the algorithm detects 8 of them, then the recall is $$8 / 10 = 0.8$$, or $$80%$$.  
```

### Results

The following table summarizes RCF's precision and recall performance using the NAB CloudWatch benchmarks.

| Dataset | Precision | Recall |
|-----|-----------|--------|
| AWS `ec2_cpu_utilization_24ae8d`  | 0.67      | 1      |
| AWS `ec2_network_in_257a54`       | 0.25      | 1      |
| AWS `ec2_disk_write_bytes_1ef3de` | 0.33      | 1      |
| AWS `rds_cpu_utilization_e47b3b`  | 1         | 1      |
| AWS `rds_cpu_utilization_cc0c53`  | 1         | 0.5    |

> Overall, RCF demonstrated strong recall, correctly detecting 7 out of 8 anomalies across the datasets. Notably, it achieved perfect recall in four of the five datasets. 

Reviewer 3: ```suggestion
Overall, RCF demonstrated strong recall, correctly detecting 7 out of 8 anomalies across the datasets. Notably, it achieved perfect recall in 4 of the 5 datasets. 
```

> RCF also maintains high precision. Most false positives occur early in the time series, before the model has accumulated sufficient historical data for accurate predictions. For example, in the following graph, the anomaly detected around April 12 at 3:14 for the `ec2_network_in_257a54` dataset is particularly noteworthy. It deviates from the previously observed pattern of uniform double spikes. Earlier patterns show the second spike declining within approximately five minutes, whereas this anomalous spike exhibits an extended decline lasting around ten minutes. 

Reviewer 3: ```suggestion
RCF also maintains high precision. Most false positives occur early in the time series, before the model has accumulated sufficient historical data to make accurate predictions. For example, in the following graph, the anomaly detected around April 12 at 3:14 for the `ec2_network_in_257a54` dataset is particularly noteworthy. It deviates from the previously observed pattern of uniform double spikes. Earlier patterns show the second spike declining within approximately 5 minutes, whereas this anomalous spike exhibits an extended decline lasting around 10 minutes. 
```

![network_zoom_in](/assets/media/blog-images/2024-12-26-Reducing-false-positives-through-algorithmic-improvements/network_zoom_in_1.png){:class="img-centered"}

> When excluding the first 20% of data as a probation period, precision improves across datasets. This adjustment highlights how RCF becomes more accurate after observing enough historical data. The results of the last 80% of the dataset are summarized in the following table.

Reviewer 3: ```suggestion
When excluding the first 20% of the data as a probation period, precision improves across datasets. This adjustment highlights how RCF becomes more accurate after observing enough historical data. The results of the last 80% of the dataset are summarized in the following table.
```

> 

Reviewer 3: Line 203, first sentence: "of the data" => "of the dataset" (to match the last sentence)?

| AWS rds_cpu_utilization_e47b3b | 1 | 1 |
| AWS rds_cpu_utilization_cc0c53 | 1 | 0.5 |

Another type of false positive occurs when detected anomalies seem valid based on the data but are in fact not labeled as anomalies. For example, in the following graph, the anomaly detected on April 15 at approximately 3:34 for the `ec2_network_in_257a54` dataset deviates from the typical pattern of paired spikes. Instead of the usual two consecutive peaks, this event shows a single spike. While this represents a departure from the established pattern, it is not labeled as an anomaly. 

![network_zoom_in_3](/assets/media/blog-images/2024-12-26-Reducing-false-positives-through-algorithmic-improvements/network_zoom_in_3.png){:class="img-centered"}

## OpenSearch 2.17 and 2.9 compared

Let's examine how OpenSearch 2.17 has improved its anomaly detection performance relative to OpenSearch 2.9. Before we present the results, let's define the following key concepts:

- **True positive (correct detection)**: An true anomalous event correctly flagged as anomalous by the algorithm.  
- **False positive (false alarm)**: A non-anomalous event incorrectly flagged as anomalous by the algorithm.  

> - **False negative (missed anomaly)**: An anomalous event that the system fails to flag as anomalous.

Reviewer 3: ```suggestion
- **False negative (missed anomaly)**: An anomalous event that the system failed to flag as anomalous.
```

| Dataset | 2.9 anomalies detected | 2.9 true positives | 2.17 anomalies detected | 2.17 true positives | Actual anomalies |
|--------------------------------------|-------------:|-------------------:|--------------:|--------------------:|-------------:|
| AWS ec2_cpu_utilization_24ae8d       |           17 |                  1 |             3 |                   2 |            2 |
| AWS ec2_network_in_257a54            |           33 |                  1 |             4 |                   1 |            1 |
| AWS ec2_disk_write_bytes_1ef3de      |           16 |                  1 |             3 |                   1 |            1 |
| AWS rds_cpu_utilization_e47b3b       |           23 |                  2 |             2 |                   2 |            2 |
| AWS rds_cpu_utilization_cc0c53       |           23 |                  1 |             1 |                   1 |            2 |
| **Totals**                           |          112 |                  6 |            13 |                   7 |            8 |

### Discussion

Version 2.17 showed an overall reduction in false alerts and an increase in true alerts compared to version 2.9. 

> OpenSearch 2.9 flagged 112 anomalies, but only 6 were true anomalies, resulting in a high false positive rate. After algorithmic improvements in OpenSearch 2.17, only 13 anomalies were flagged, with 7 being true anomalies. This reflects a significant increase in precision, from 5.4% (6/112) to 53.8% (7/13) and increase in recall from 75% (6/8) to 87.5% (7/8).

Reviewer 3: ```suggestion
OpenSearch 2.9 flagged 112 anomalies, but only 6 were true anomalies, resulting in a high false positive rate. After algorithmic improvements in OpenSearch 2.17, only 13 anomalies were flagged, with 7 being true anomalies. This reflects a significant increase in precision, from 5.4% (6/112) to 53.8% (7/13), and an increase in recall from 75% (6/8) to 87.5% (7/8).
```

> In comparison, OpenSearch 2.9 produced 106 false positives, while OpenSearch 2.17 produced just 6—--representing a **94.3% reduction in false positives**. This percent reduction can be calculated using the following formula:

Reviewer 3: ```suggestion
In comparison, OpenSearch 2.9 produced 106 false positives, while OpenSearch 2.17 produced just 6---representing a **94.3% reduction in false positives**. This percent reduction can be calculated using the following formula:
```

$$
\text{Percent Reduction in FP} 
= \frac{\text{Old FP} - \text{New FP}}{\text{Old FP}} \times 100\%
= \frac{106 - 6}{106} \times 100\% = 94.3\%
$$

> Similarly, OpenSearch 2.9 missed 2 real anomalies, whereas OpenSearch 2.17 missed just 1--—representing a **50% reduction in false negatives**, computed as follows:

Reviewer 3: ```suggestion
Similarly, OpenSearch 2.9 missed 2 real anomalies, whereas OpenSearch 2.17 missed just 1---representing a **50% reduction in false negatives**, computed as follows:
```
